ocaml/libs/tracing/tracing.ml

```diff
@@ -71,11 +90,92 @@ module Spans = struct
 end
 
 module Tracer = struct
-  let start ~name ~parent : (Span.t option, exn) result =
-    let span = Span.start ~name ~parent () in
-    Spans.add_to_spans ~span ; Ok (Some span)
+  type t = {name: string; provider: provider_config_t ref}
+
+  let create ~name ~provider = {name; provider}
+
+  let no_op =
+    let provider =
+      ref
+        {
+          name_label= ""
+        ; tags= []
+        ; endpoints= []
+        ; filters= []
+        ; processors= []
+        ; enabled= false
+        }
+    in
+    {name= ""; provider}
+
+  let start ~tracer:t ~name ~parent : (Span.t option, exn) result =
+    let provider = !(t.provider) in
+    (* Do not start span if the TracerProvider is diabled*)
+    if not provider.enabled then
+      Ok None
+    else
+      let tags = provider.tags in
+      let span = Span.start ~tags ~name ~parent () in
+      Spans.add_to_spans ~span ; Ok (Some span)
 
   let finish (span : Span.t option) : (unit, exn) result =
     let _span = Option.map (fun span -> Span.finish ~span ()) span in
     Ok ()
 end
+
+module TracerProvider = struct
+  type t = {tracers: Tracer.t list; config: provider_config_t}
+
+  let get_tracer ~provider:t ~name =
+    match
+      List.filter (fun (tracer : Tracer.t) -> tracer.name = name) t.tracers
```

Comment:
Are we going to look up tracers like this often? How many tracers do you normally have?
If it is a small number, like 1 or 2 then the List may be better, if you expect to have lots (hundreds) then a String Map might be better (O(log n) vs O(n) lookup time).

Reply:
@jameshensmancitrix

Reply:
Will delete as not used